text/0000-llvm-asm.md

```diff
@@ -0,0 +1,73 @@
+- Feature Name: `llvm_asm`
+- Start Date: 2019-12-31
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Deprecate the existing `asm!` macro and provide an identical one called
+`llvm_asm!`. The feature gate is also renamed from `asm` to `llvm_asm`.
+
+Unlike `asm!`, `llvm_asm!` is not intended to ever become stable.
```

Comment:
This sounds a bit too strong, e.g, if we ever add a `#[cfg(codegen_backed = "...")]` macro to detect the code generation backend, then using `llvm_asm!` behind a `#[cfg(codegen_backend = "llvm")]` might make sense. 

```suggestion
Unlike `asm!`, `llvm_asm!` is currently perma unstable.
```

Also, if the feature is intended to be perma-unstable, at least for the time being, it might make sense to name the feature gate: `feature(rustc_llvm_asm)`.

Reply:
Last that I checked (which was a while ago, mind), there are more objections to stabilizing the existing `asm!` macro than just its coupling to LLVM. Furthermore, if the proto-RFC for stable ASM were accepted and implemented it would be sufficiently different from the existing ASM support (both syntactically and semantically) that I feel people would be reluctant to stabilize the old version, if only for the sake of consistency.

This isn't to say anything about whether a stable `llvm_asm!` might be desirable; only that if it is desirable, it would likely not be served by merely stabilizing the existing ASM macro. I would expect such a discussion to be the purview of a totally separate RFC, although if the possibility isn't completely remote then that could suggest leaving the name `llvm_asm!` free for future use. I do like the idea of prepending "rustc" to things which are not intended to ever be stabilized... but really the "rustc" prefix should only be used for perma-unstable compiler implementation details, and unless rustc is using `asm!` (and I don't see why it would), I'd much prefer planning to remove the original `asm!` sometime after a replacement has been stabilized.

Reply:
My position is that the new `asm!` should be designed to provide substantially all of the functionality of the old one – and eventually more, such as an equivalent of C [`asm goto`](https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#GotoLabels), which is supported by LLVM but not currently exposed in Rust.\*  After all... well, I don't want to downplay the complexity of implementing inline assembly in the first place.  But since we *are* planning to do that, the question is how much additional complexity would be required to get from "basic support" to "full parity".  As I see it, most of that complexity comes from two sources:

1. The large amount of functionality supported by the assembler itself (e.g. directives, preprocessor)... but we're aiming to target existing assemblers, not create our own.

2. The [smorgasbord of arch-dependent constraint codes](https://gcc.gnu.org/onlinedocs/gcc/Machine-Constraints.html) supported by GCC... but LLVM only supports a small fraction of them anyway, and we only need to match LLVM's support.

So it should be feasible to reach pretty much full parity.  At that point, `llvm_asm!` will be redundant, and I think it's fine to plan to remove it.  That said, there would be no urgent *need* to remove it; it should be possible to have a long deprecation period, and there might even be a case for keeping it indefinitely for experimentation purposes... although that would have to be weighed against the compiler maintenance burden.

\* For anyone who hasn't been following discussions, that doesn't mean adding goto to Rust; we just need some way to express the concept of jumping out of asm blocks, which would most likely resemble a match statement.

Reply:
The only thing we need to unblock progress on the new inliner assembler is to rename the `asm!` macro to `llvm_asm!`, although as exposed here it also makes sense to rename the feature gate. 

While I mostly agree with what you said, we don't really need to make any commitments right now about the future of `llvm_asm!`, so we can just punt that discussion.